### PR TITLE
feat(devices): proposal for reflecting new device structure

### PR DESCRIPTION
**Pull request by JavierSab**
## TL;DR

This PR makes viable the homonymouis [#27](https://github.com/qilimanjaro-tech/qili-global-quantum-service/pull/354) from QGQS.

##  The  refactor: user perspective

Its main objective was accomodating for the new device structure. This has been done in a way that is compatible with the old one, making it viable to distribute this qiboconnection version before deploying the backend changes.

The users, however, will see some modifications on how the device information is represented.

With the currently main qiboconnection code, the structure they perceive is:
```bash
<Devices[3]:
{
  "device_id": 1,
  "device_name": "Radagast simulator",
  "status": "online",
  "availability": "available",
  "characteristics": {
    "type": "simulator",
    "cpu": "Intel Core i9-9900K @ 16x 5GHz",
    "gpu": "NVIDIA GeForce RTX 3090",
    "os": "Ubuntu 20.04 focal",
    "kernel": "x86_64 Linux 5.4.0-80-generic",
    "ram": "64185MiB"
  }
}
{
  "device_id": 9,
  "device_name": "Galadriel Qblox rack",
  "status": "online",
  "availability": "available",
  "characteristics": {
    "type": "quantum"
  },
  "calibration_details": {
    "t1": 10,
    "frequency": 988
  }
}
{
  "device_id": 16,
  "device_name": "Saruman",
  "status": "maintenance",
  "availability": "available",
  "characteristics": {
    "type": "quantum"
  },
  "calibration_details": {
    "t1": 10,
    "frequency": 988
  }
}
```

With the new changes from this branch, attacking the same backend this is what they will see. Notice that `device_id`and `device_name` have become `id` and `name`, and that the structure across devices has been standarized, having as result some nulls in the shown dictionaries.
```bash
[{
  "id": 1,
  "name": "Radagast simulator",
  "status": "online",
  "type": null,
  "availability": "available",
  "number_pending_jobs": null,
  "slurm_partition": "radagast_qaas",
  "static_features": null,
  "dynamic_features": null,
  "channel_id": 2,
  "characteristics": {
    "type": "simulator",
    "cpu": "Intel Core i9-9900K @ 16x 5GHz",
    "gpu": "NVIDIA GeForce RTX 3090",
    "os": "Ubuntu 20.04 focal",
    "kernel": "x86_64 Linux 5.4.0-80-generic",
    "ram": "64185MiB"
  },
  "str": "<Device: device_id=1, device_name='Radagast simulator', status='online', availability='available'>"
},
{
  "id": 9,
  "name": "Galadriel Qblox rack",
  "status": "online",
  "type": null,
  "availability": "available",
  "number_pending_jobs": null,
  "slurm_partition": "galadriel_qaas",
  "static_features": null,
  "dynamic_features": null,
  "channel_id": 2,
  "characteristics": {
    "type": "quantum"
  },
  "calibration_details": {
    "t1": 10,
    "frequency": 988
  },
  "str": "<Device: device_id=9, device_name='Galadriel Qblox rack', status='online', availability='available'>"
},
{
  "id": 16,
  "name": "Saruman",
  "status": "maintenance",
  "type": null,
  "availability": "available",
  "number_pending_jobs": null,
  "slurm_partition": "saruman_qaas",
  "static_features": null,
  "dynamic_features": null,
  "channel_id": 2,
  "characteristics": {
    "type": "quantum"
  },
  "calibration_details": {
    "t1": 10,
    "frequency": 988
  },
  "str": "<Device: device_id=16, device_name='Saruman', status='maintenance', availability='available'>"
},
]
```

Once we update the backend, the structure that they should be seeing is the following one:

```bash
<Devices[3]:
[{
  "id": 10,
  "name": "offline",
  "status": "offline",
  "type": "device",
  "availability": "available",
  "number_pending_jobs": null,
  "slurm_partition": "",
  "static_features": {},
  "dynamic_features": {},
  "str": "<Device: device_id=10, device_name='offline', status='offline', availability='available'>"
},
{
  "id": 5,
  "name": "qili_quantum_legacy",
  "status": "online",
  "type": "device",
  "availability": "available",
  "number_pending_jobs": null,
  "slurm_partition": "qili_quantum_slurm_qaas",
  "static_features": {},
  "dynamic_features": {},
  "str": "<Device: device_id=5, device_name='qili_quantum_legacy', status='online', availability='available'>"
},
{
  "id": 4,
  "name": "qili_simulator_legacy",
  "status": "online",
  "type": "device",
  "availability": "available",
  "number_pending_jobs": null,
  "slurm_partition": "qili_simulator_slurm_qaas",
  "static_features": {
    "graph": {
      "links": [
        {
          "source": 0,
          "target": 2
        },
        {
          "source": 1,
          "target": 2
        },
        {
          "source": 3,
          "target": 2
        },
        {
          "source": 4,
          "target": 2
        }
      ],
      "nodes": [
        {
          "id": 0,
          "x": 0,
          "y": 0
        },
        {
          "id": 1,
          "x": 0,
          "y": 2
        },
        {
          "id": 2,
          "x": 1,
          "y": 1
        },
        {
          "id": 3,
          "x": 2,
          "y": 0
        },
        {
          "id": 4,
          "x": 2,
          "y": 2
        }
      ]
    },
    "type": "analog"
  },
  "dynamic_features": {},
  "str": "<Device: device_id=4, device_name='qili_simulator_legacy', status='online', availability='available'>"
},
]
```

##  The  refactor: developer perspective

In big terms, and in the same direction of the respective qgqs refactor, this PR eliminates all device differenciation. If we want to hide this from the user, an abstract factory could be implemented so that we prepare methods for getting the data, but it was agreed to try the simpler approach before complicating the logic.

Previously, the structural validation was done across different classes. Here we reduce it to two: DeviceInput and Device.

[DeviceInput](https://github.com/qilimanjaro-tech/qiboconnection/blob/0feed7ac22f844844abcf9cb588db74be4d2cb5b/src/qiboconnection/typings/devices/device_input.py) is the one that was supposed to validate the incoming dictionaries. Its input is the result from json-loading whatever comes from the api. The changes' key-points are:
* We only require the devices to have an `id`, a `name` and a `status`. If instead the old names for those fields were present in the received dictionaries, we parse them here: https://github.com/qilimanjaro-tech/qiboconnection/blob/0feed7ac22f844844abcf9cb588db74be4d2cb5b/src/qiboconnection/typings/devices/device_input.py#L50-L61
* We expect the devices to also possibly present a number of attributes, that we present on this class. If not present, they will appear as a None.
* We allow for additional fields to be present, and they will be shown to the user. This is what the `from_kwargs` function takes care of.

[Device](https://github.com/qilimanjaro-tech/qiboconnection/blob/0feed7ac22f844844abcf9cb588db74be4d2cb5b/src/qiboconnection/models/devices/device.py) cares for the logic of presenting that information for the user. For this:
* It assumes a device input will be handed for it to be built. Therefore, it does not validate inputs, instead declaring as attributes all the fields from the provided dictionary (with the format "_fieldname"). 
* Then, it presents as **readonly properties** the fields that we would expect a user would want to access. More information may be present on the device, but in case they want to iterate and search for specific fields, or check the class documentation, this might be a good way where documenting said fields.
* Finally, it defines a `to_dict()` method for making easier the class' representation. It iterates over all single-underscore prefixed fields and presents them as a json.

The difference between devices has been "hidden" in plain-sight in 3 core places: `type`, `static_properties` and `dynamic_properties`. 
* The first one will have a string, that is an enum in public but that we may not want to force to be an enum here. 
* The second and third ones are dicts supposed to hold the stuff that is deffinitive and the stuff that changes from a device, and its content is what will depend on the type of device. A real quantum device might have its graph, qbit technology, etc handed on its static_properties while the calibration information is on the dynamic_properties, and a simulator might have information about its hardware, type of simulation (algebraic, cqed, etc) on its static_properties, and maybe some performance metrics or library versions on the dynamic_properties.

### CODE REVIEW

**Comment by JavierSab** [comment]
After implementing the `expand: bool` kwarg to the `__str__()`/`__repr__()`/`to_dict()`/`toJSON()` functions, listing looks like:

```bash
<Devices[3]:[
  {
    "type": null,
    "number_pending_jobs": null,
    "name": "Radagast simulator",
    "id": 1,
    "status": "online"
  },
  {
    "type": null,
    "number_pending_jobs": null,
    "name": "Galadriel Qblox rack",
    "id": 9,
    "status": "online"
  },
  {
    "type": null,
    "number_pending_jobs": null,
    "name": "Saruman",
    "id": 16,
    "status": "maintenance"
  }
]>
```

**Review by ziiiki — APPROVED** [review]

**Review by DZPM — APPROVED** [review]
lgtm
